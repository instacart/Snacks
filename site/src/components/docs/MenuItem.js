import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

  <Fragment>
    MenuItem should always be wrapped in parent Menu or DropdownMenu. And onClick should be accessed through the parent's onSelect callback.
  </Fragment>

)
