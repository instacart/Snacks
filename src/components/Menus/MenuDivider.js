import React      from 'react'
import PropTypes  from 'prop-types'
import { colors } from '../../styles'
import spacing    from '../../styles/spacing'

const styles = {
  backgroundColor: colors.GRAY_88,
  width: '100%',
  ...spacing.MARGIN_Y_XS,
  marginLeft: 0,
  marginRight: 0,
  height: 1,
  border: 'none'
}

const MenuDivider = (props) => {
  const { style } = props

  return (
    <hr style={{...styles, ...style}}/>
  )
}

MenuDivider.propTypes = {
  /** Override default styles */
  style: PropTypes.shape({})
}

export default MenuDivider
