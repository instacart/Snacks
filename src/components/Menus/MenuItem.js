import React from 'react'
import PropTypes from 'prop-types'
import Radium from 'radium'
import { colors } from '../../styles'
import Icon from '../Icon/Icon'
import spacing from '../../styles/spacing'

/* eslint jsx-a11y/no-noninteractive-tabindex: 0 */

const styles = {
  root: {
    backgroundColor: '#ffffff',
    ...spacing.PADDING_Y_SM,
    paddingRight: 0,
    paddingLeft: 0,
    ':hover': {
      cursor: 'pointer',
      backgroundColor: colors.GRAY_97,
    },
    ':focus': {
      backgroundColor: colors.GRAY_97,
    },
  },
  focus: {
    backgroundColor: colors.GRAY_97,
  },
  disabled: {
    color: colors.GRAY_74,
    backgroundColor: '#FFF',
    pointerEvents: 'none',
  },
  menuitem: {
    display: 'table',
  },
  iconContainer: {
    display: 'table-cell',
    verticalAlign: 'middle',
    paddingLeft: '16px',
  },
  label: {
    ...spacing.PADDING_X_SM,
    paddingTop: 0,
    paddingBottom: 0,
    whiteSpace: 'nowrap',
  },
  leftIconStyles: {
    fontSize: '22px',
  },
}

@Radium
class MenuItem extends React.Component {
  static propTypes = {
    /** Completely override the MenuItem rendering and create a custom MenuItem */
    children: PropTypes.node,
    /** Disable the MenuItem */
    disabled: PropTypes.bool,
    /** Focus the MenuItem */
    focus: PropTypes.bool,
    /** Index of MenuItem within Menu. Used for currentIndex */
    index: PropTypes.number,
    /** Text for the menu item */
    label: PropTypes.string.isRequired,
    /** Override styles of label */
    labelStyles: PropTypes.shape({}),
    /** Icon name or Icon component displayed left of label */
    leftIcon: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    /** Override styles for leftIcon */
    leftIconStyles: PropTypes.shape({}),
    /** Callback function fired when the menu item is click. Overriden by parent Menu or DropdownMenu */
    _onClick: PropTypes.func,
    /** Callback function fired when the menu item is focused. */
    onFocus: PropTypes.func,
    /** Used by menu to keep track of current focus index. */
    onMenuItemFocus: PropTypes.func,
    /** Whether or not to prevent default when menu item is clicked */
    preventDefault: PropTypes.bool,
    /** Role HTML attribute */
    role: PropTypes.string,
    /** Customize style of MenuItem */
    style: PropTypes.shape({}),
    /** Override tabIndex property */
    tabIndex: PropTypes.number,
    /** Whether or not to use use the tabIndex HTML attribute */
    useTabIndex: PropTypes.bool,
    /** Underlying value. Also, passed into _onClick function */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]).isRequired,
  }

  static defaultProps = {
    disabled: false,
    role: 'menuitem',
    useTabIndex: true,
    focus: false,
    preventDefault: true,
  }

  static isSnacksMenuItem = true

  componentDidUpdate(prevProps) {
    if (!prevProps.focus && this.props.focus) {
      setTimeout(() => {
        // Checking the ref exists in case this component
        // unmounts before the callback runs.
        this.menuItem && this.menuItem.focus()
      }, 0)
    }
  }

  handleClick = e => {
    const { disabled, index, _onClick, value, label, preventDefault } = this.props
    preventDefault && e.preventDefault()

    const option = { value, label }
    !disabled && _onClick && _onClick(e, option, index)
  }

  handleKeyDown = event => {
    switch (event.key) {
      case 'Enter':
        this.props.preventDefault && event.preventDefault()
        this.handleClick(event)
    }
  }

  handleFocus = event => {
    const { onFocus, onMenuItemFocus, index } = this.props
    onMenuItemFocus && onMenuItemFocus(index)
    onFocus && onFocus(index, event)
  }

  renderLeftIcon() {
    const { leftIcon, leftIconStyles } = this.props
    if (!leftIcon) {
      return
    }

    const iconComponent =
      typeof leftIcon === 'string' ? (
        <Icon name={leftIcon} style={[styles.leftIconStyles, leftIconStyles]} />
      ) : (
        leftIcon
      )

    return (
      <div style={styles.iconContainer}>
        <div style={{ display: 'flex' }}>{iconComponent}</div>
      </div>
    )
  }

  renderMenuItem() {
    const { children, label, labelStyles } = this.props

    if (children) {
      return children
    }

    return (
      <div style={styles.menuitem}>
        {this.renderLeftIcon()}
        <div style={[styles.label, labelStyles]}>{label}</div>
      </div>
    )
  }

  render() {
    const { disabled, role, style, tabIndex, useTabIndex } = this.props

    return (
      <div
        ref={node => (this.menuItem = node)}
        role={role}
        style={[styles.root, style, disabled && styles.disabled]}
        onClick={this.handleClick}
        onFocus={this.handleFocus}
        tabIndex={useTabIndex && !disabled ? tabIndex || 0 : undefined}
        onKeyDown={this.handleKeyDown}
      >
        {this.renderMenuItem()}
      </div>
    )
  }
}

export default MenuItem
