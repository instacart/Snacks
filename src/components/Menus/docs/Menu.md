Menu Example:

```js

function onSelect(e, foo) {
  console.log(foo)
}

<div style={{display: 'flex'}}>
  <div style={{marginRight: '30px'}}>
    <Menu onSelect={onSelect}>
      <MenuItem label="First" value="First" />
      <MenuItem label="Second" value="Second" />
      <MenuItem label="Third" value="Third" disabled />
      <MenuItem label="Fourth" value="Fourth"/>
      <MenuDivider />
      <MenuItem label="Fifth" value="Fifth" />
    </Menu>
  </div>

  <div>
    <Menu>
      <MenuItem label="Share via Email" value="email" leftIcon="emailFilled" labelStyles={{color: 'green'}}/>
      <MenuItem label="Share via Facebook" value="facebook" leftIcon={<Icon name="facebookFilled" style={{color: '#3c5a99', fontSize: '22px'}} />} />
      <MenuItem label="Share via Twitter" value="twitter" leftIcon="twitterFilled" leftIconStyles={{color: '#1da1f2'}}/>
    </Menu>
  </div>
</div>
```
