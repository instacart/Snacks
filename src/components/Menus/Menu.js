import React             from 'react'
import PropTypes         from 'prop-types'
import MenuItem          from './MenuItem'
import getRuntimeType    from '../../utils/getRuntimeType'

const styles = {
  borderRadius: '4px',
  backgroundColor: '#ffffff',
  boxShadow: '0 3px 10px 0 rgba(0, 0, 0, 0.16), 0 3px 10px 0 rgba(0, 0, 0, 0.23)',
  display: 'inline-block',
  padding: '8px 0',
  overflowY: 'auto',
  userSelect: 'none',
  maxHeight: '500px',
  outline: 'none'
}

class Menu extends React.Component {
  static propTypes = {
    /** aria-hidden HTML attribute */
    ariaHidden: PropTypes.bool,
    /** MenuItems or Divider */
    children: PropTypes.node.isRequired,
    /** Callback function fired on key down */
    onKeyDown: PropTypes.func,
    /** Callback function fired when a MenuItem is selected */
    onSelect: PropTypes.func,
    /** Callback function fired when a Menu is blurred */
    onBlur: PropTypes.func,
    /** Role HTML attribute */
    role: PropTypes.string,
    /** Customize style of menu parent */
    style: PropTypes.shape({})
  }

  static defaultProps = {
    role: 'menu',
    ariaHidden: false,
    style: {},
    onSelect: () => {} // eslint-disable-line no-empty-function
  }

  state = {
    currentTabIndex: null
  }

  handleBlur = (event) => {
    const { onBlur } = this.props
    const currentTarget = event.currentTarget

    setTimeout( () => {
      if (!currentTarget.contains(document.activeElement)) {
        this.setState({currentTabIndex: null}, () => {
          onBlur && onBlur(event)
        })
      }
    }, 0)
  }

  handleKeyDown = (event) => {
    const { onKeyDown } = this.props

    switch(event.key) {
      case 'ArrowDown': {
        event.preventDefault()
        this.incrementTabIndex()
        break
      }
      case 'ArrowUp': {
        event.preventDefault()
        this.decrementTabIndex()
        break
      }
    }

    onKeyDown && onKeyDown(event)
  }

  handleMenuItemFocus = (index) => {
    this.setState({currentTabIndex: index})
  }

  blur() {
    this.menu.blur()
  }

  focus() {
    this.menu.focus()
  }

  incrementTabIndex() {
    this.setState({currentTabIndex: this.nextValidTabIndex()})
  }

  decrementTabIndex() {
    this.setState({currentTabIndex: this.prevValidTabIndex()})
  }

  nextValidTabIndex() {
    const { currentTabIndex } = this.state
    const menuItemChildren = this.getMenuItemChildren()
    const maxIndex = menuItemChildren.length - 1
    const newIndex = currentTabIndex !== null ? currentTabIndex+1 : 0

    for (let index = newIndex; index <= maxIndex; index++) {
      const menuItem = menuItemChildren[index]

      if (this.menuItemIsValid(menuItem)) {
        return index
      }
    }

    return currentTabIndex
  }

  prevValidTabIndex() {
    const { currentTabIndex } = this.state
    const menuItemChildren = this.getMenuItemChildren()

    for (let index = (currentTabIndex-1); index >= 0; index--) {
      const menuItem = menuItemChildren[index]

      if (this.menuItemIsValid(menuItem)) {
        return index
      }
    }

    return currentTabIndex
  }

  menuItemIsValid(menuItem) {
    return menuItem && !menuItem.props.disabled
  }

  getMenuItemChildren() {
    const { children } = this.props
    const menuItemChildren = []

    React.Children.map(children, (child) => {
      if (child.type && child.type === getRuntimeType(MenuItem)) {
        menuItemChildren.push(child)
      }
    })

    return menuItemChildren
  }

  renderChildren() {
    const { children, onSelect } = this.props
    const { currentTabIndex } = this.state
    let index = 0

    return React.Children.map(children, (child) => {
      if (!React.isValidElement(child)) {
        throw 'Passing invalid element to Menu'
      } else if (child.type && child.type === getRuntimeType(MenuItem)) {
        const component = React.cloneElement(child, {
          index: index,
          focus: currentTabIndex === index,
          _onClick: onSelect,
          onMenuItemFocus: this.handleMenuItemFocus
        })
        index += 1
        return component
      }

      return child
    })
  }

  render() {
    const {
      ariaHidden,
      style,
      role,
    } = this.props

    const mergedStyles = {
      ...styles,
      ...style
    }

    return (
      <div
        ref={(node) => this.menu = node}
        role={role}
        style={mergedStyles}
        onKeyDown={this.handleKeyDown}
        onBlur={this.handleBlur}
        aria-hidden={ariaHidden}
        tabIndex={-1}
      >
        { this.renderChildren() }
      </div>
    )
  }
}

export default Menu
