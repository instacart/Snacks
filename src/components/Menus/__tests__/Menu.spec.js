import React from 'react'
import renderer from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount } from 'enzyme'
import { spy } from 'sinon'
import Menu from '../Menu'
import MenuItem from '../MenuItem'
import MenuDivider from '../MenuDivider'

it('renders Menu with MenuItems correctly', () => {
  const tree = renderer
    .create(
      <StyleRoot>
        <div>
          <Menu>
            <MenuItem label="First" value="First" />
            <MenuItem label="Second" value="Second" />
            <MenuItem label="Third" value="Third" />
            <MenuDivider />
            <MenuItem label="Fourth" value="Fourth" />
          </Menu>
        </div>
      </StyleRoot>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should render all of MenuItem children', () => {
  const wrapper = mount(
    <Menu>
      <MenuItem label="First" value="First" />
      <MenuItem label="Second" value="Second" />
      <MenuDivider />
      <MenuItem label="third" value="Third" />
    </Menu>
  )
  expect(wrapper.find(MenuItem).length).toBe(3)
  expect(wrapper.find(MenuDivider).length).toBe(1)
})

it('should fire onSelect when a MenuItem is selected', () => {
  const onSelect = spy()
  const wrapper = mount(
    <StyleRoot>
      <div>
        <Menu onSelect={onSelect}>
          <MenuItem label="First" value="First" />
          <MenuItem label="Second" value="Second" />
        </Menu>
      </div>
    </StyleRoot>
  )

  wrapper
    .find(MenuItem)
    .first()
    .simulate('click')
  expect(onSelect.calledOnce).toBe(true)
})
