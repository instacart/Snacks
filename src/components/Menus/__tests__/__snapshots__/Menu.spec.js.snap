// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Menu with MenuItems correctly 1`] = `
<div
  data-radium={true}
>
  <div
    data-radium={true}
  >
    <div
      aria-hidden={false}
      onBlur={[Function]}
      onKeyDown={[Function]}
      role="menu"
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": "4px",
          "boxShadow": "0 3px 10px 0 rgba(0, 0, 0, 0.16), 0 3px 10px 0 rgba(0, 0, 0, 0.23)",
          "display": "inline-block",
          "maxHeight": "500px",
          "outline": "none",
          "overflowY": "auto",
          "paddingBottom": 8,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 8,
          "userSelect": "none",
        }
      }
      tabIndex={-1}
    >
      <div
        data-radium={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "backgroundColor": "#ffffff",
            "paddingBottom": 16,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 16,
          }
        }
        tabIndex={0}
      >
        <div
          data-radium={true}
          style={
            Object {
              "display": "table",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "whiteSpace": "nowrap",
              }
            }
          >
            First
          </div>
        </div>
      </div>
      <div
        data-radium={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "backgroundColor": "#ffffff",
            "paddingBottom": 16,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 16,
          }
        }
        tabIndex={0}
      >
        <div
          data-radium={true}
          style={
            Object {
              "display": "table",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "whiteSpace": "nowrap",
              }
            }
          >
            Second
          </div>
        </div>
      </div>
      <div
        data-radium={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "backgroundColor": "#ffffff",
            "paddingBottom": 16,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 16,
          }
        }
        tabIndex={0}
      >
        <div
          data-radium={true}
          style={
            Object {
              "display": "table",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "whiteSpace": "nowrap",
              }
            }
          >
            Third
          </div>
        </div>
      </div>
      <hr
        style={
          Object {
            "backgroundColor": "#E0E0E0",
            "border": "none",
            "height": 1,
            "marginBottom": 8,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 8,
            "width": "100%",
          }
        }
      />
      <div
        data-radium={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "backgroundColor": "#ffffff",
            "paddingBottom": 16,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 16,
          }
        }
        tabIndex={0}
      >
        <div
          data-radium={true}
          style={
            Object {
              "display": "table",
            }
          }
        >
          <div
            data-radium={true}
            style={
              Object {
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "whiteSpace": "nowrap",
              }
            }
          >
            Fourth
          </div>
        </div>
      </div>
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
